Containerized workflow

ozkanonur opened the pull request:
- [x] Create the containers for Anitya
- - [x] web
- - [x] libraries.io consumer
- [x] Create a container for RabbitMQ
- [x] Create a container for Postgres DB
- [x] Anitya containers should be able to connect to RabbitMQ and Postgres DB containers
- [x] Update contribution guide

fixes #936

ozkanonur commented:
Since I am using mobile internet (because I am on vacation), I do development on remote servers because building container images use too much internet. But, it's not really comfortable the way working like that. I want to continue this task after 1-2 weeks if it's not a problem.

abitrolly commented:
@ozkanonur I would use GitLab CI/CD. You still need to test them and make the build automated, so why not to start directly from CI/CD.

ozkanonur commented:
> @ozkanonur I would use GitLab CI/CD. You still need to test them and make the build automated, so why not to start directly from CI/CD.

I guess good idea. I will try that way too.

Zlopez commented:
@ozkanonur Take your time and enjoy your vacation. Your work is really appreciated.

Zlopez commented:
@abitrolly We can use Zuul for this, it's already available on Anitya.

ozkanonur commented:
> @abitrolly We can use Zuul for this, it's already available on Anitya.

I thought @abitrolly tried to tell me another way to avoid using the internet on image buildings. :smile:  Debugging/Developing on Gitlab CD/CI is actually faster than working on free tier AWS servers. :joy: But still not as easy and comfortable as host machine :)

abitrolly commented:
Yea, vacation is sacred, so stop reading this! :D

@ozkanonur why didn't you start with https://github.com/fedora-infra/anitya/blob/master/Dockerfile ? I believe it already does like more that a half of the work by reusing Ansible playbooks.

ozkanonur commented:
> Yea, vacation is sacred, so stop reading this! :D
> 
> @ozkanonur why didn't you start with https://github.com/fedora-infra/anitya/blob/master/Dockerfile ? I believe it already does like more that a half of the work by reusing Ansible playbooks.

I know I see it but I wanted to keep containers as lightweight and simple as possible.

Btw working on vacation feels a little sweet :P
![E6qhlzOXsAQ_Nby](https://user-images.githubusercontent.com/39852038/128680935-65ec54cb-7c42-47b6-b11d-36f0dcf972dd.jpeg)

abitrolly commented:
> I know I see it but I wanted to keep containers as lightweight and simple as possible.

Yes. I didn't care about container size much as long as I can speed up development, but it would be interesting to see the metrics with comparison too.

ozkanonur commented:
> @ozkanonur great view. Kind of envy you. :D Which mountains are there?

The place is called "Limonsuyu Yaylasi" from Trabzon/Turkey. Here we are above 3500m of the town. I uploaded the least beautiful picture to not envy you guys actually :D I am sorry for that :D

I can do the comparison between containers with and without ansible once I get back into my home.

ozkanonur commented:
Is RabbitMQ being used while developing the project ? Because I don't see any amqp connection string anywhere in the project except [here](https://github.com/fedora-infra/anitya/blob/7e515a4030167f45272512a9d15dafdbe6e17f82/files/config.toml.sample#L5).

Zlopez commented:
> Is RabbitMQ being used while developing the project ? Because I don't see any amqp connection string anywhere in the project except [here](https://github.com/fedora-infra/anitya/blob/7e515a4030167f45272512a9d15dafdbe6e17f82/files/config.toml.sample#L5).
> 
> So, not sure where to put connection string for RabbitMQ inside the docker.

This would be probably read from default fedora messaging configuration. You can find the dev configuration used by current dev env of Anitya [here](https://github.com/fedora-infra/anitya/blob/7015f20f9901b2c40842e2adc7e8ba5489975fe8/ansible/roles/anitya-dev/files/anitya.toml).

ozkanonur commented:
Can you check if it works as it should be ? Running `make up` should provide containers for web and consumer, then you can enter into containers with `make bash-web` and `make bash-consumer` and start the application. I didn't have any issue running the projects, but it's better if you test it. If everything works as expected, I will update the contribution guides and check podman compatibility. @Zlopez

Zlopez commented:
I will test it. I just need to layer podman-compose to my Silverblue installation to be able to run it :-)

Zlopez commented:
I see some things about the current state:

1. `make up` failed to get images for rabbitmq and postgres, I got to choose from 4 sources and got `Error reading manifest`, I'm not sure which source should I use
2. How is the postgres initialized, the vagrant allows you to use latest production database backup to initialize the database, otherwise you are starting with empty database, which is not great
3. `make bash-web` doesn't work with podman-compose, because the podman-compose doesn't know `exec` `podman-compose: error: argument command: invalid choice: 'exec' (choose from 'help', 'version', 'pull', 'push', 'build', 'up', 'down', 'ps', 'run', 'start', 'stop', 'restart', 'logs')`

So I wasn't able to do much with this :-/

ozkanonur commented:
> I found some issues in the current state:
> 
> 1. `make up` failed to get images for rabbitmq and postgres, I got to choose from 4 sources and got `Error reading manifest`, I'm not sure which source should I use
> 2. How is the postgres initialized, the vagrant allows you to use latest production database backup to initialize the database, otherwise you are starting with empty database, which is not great
> 3. `make bash-web` doesn't work with podman-compose, because the podman-compose doesn't know `exec` `podman-compose: error: argument command: invalid choice: 'exec' (choose from 'help', 'version', 'pull', 'push', 'build', 'up', 'down', 'ps', 'run', 'start', 'stop', 'restart', 'logs')`
> 
> So I wasn't able to do much with this :-/

1- Hmm thats pretty weird. I will check that one.
2- Thanks for letting me know, I will migrate the db when its first time creating container.
3- podman-compose doesnt support many of docker-compose features and its not very stable. Since podman has their own compabality with docker-compose, you dont need to use podman-compose no more.

Zlopez commented:
> > I found some issues in the current state:
> > 
> > 1. `make up` failed to get images for rabbitmq and postgres, I got to choose from 4 sources and got `Error reading manifest`, I'm not sure which source should I use
> > 2. How is the postgres initialized, the vagrant allows you to use latest production database backup to initialize the database, otherwise you are starting with empty database, which is not great
> > 3. `make bash-web` doesn't work with podman-compose, because the podman-compose doesn't know `exec` `podman-compose: error: argument command: invalid choice: 'exec' (choose from 'help', 'version', 'pull', 'push', 'build', 'up', 'down', 'ps', 'run', 'start', 'stop', 'restart', 'logs')`
> > 
> > So I wasn't able to do much with this :-/
> 
> 1- Hmm thats pretty weird. I will check that one.
> 2- Thanks for letting me know, I will migrate the db when its first time creating container.
> 3- podman-compose doesnt support many of docker-compose features and its not very stable. Since podman has their own compabality with docker-compose, you dont need to use podman-compose no more.

For the database it should be good to let user choose, to have empty database is sometime useful.

When I left `docker-compose` in Makefile, it just said that `docker-compose` is not a command and the whole `make up` failed immediately. After I changed it to `podman-compose` it started doing something.

ozkanonur commented:
> > > I found some issues in the current state:
> > > 
> > > 1. `make up` failed to get images for rabbitmq and postgres, I got to choose from 4 sources and got `Error reading manifest`, I'm not sure which source should I use
> > > 2. How is the postgres initialized, the vagrant allows you to use latest production database backup to initialize the database, otherwise you are starting with empty database, which is not great
> > > 3. `make bash-web` doesn't work with podman-compose, because the podman-compose doesn't know `exec` `podman-compose: error: argument command: invalid choice: 'exec' (choose from 'help', 'version', 'pull', 'push', 'build', 'up', 'down', 'ps', 'run', 'start', 'stop', 'restart', 'logs')`
> > > 
> > > So I wasn't able to do much with this :-/
> > 
> > 
> > 1- Hmm thats pretty weird. I will check that one.
> > 2- Thanks for letting me know, I will migrate the db when its first time creating container.
> > 3- podman-compose doesnt support many of docker-compose features and its not very stable. Since podman has their own compabality with docker-compose, you dont need to use podman-compose no more.
> 
> For the database it should be good to let user choose, to have empty database is sometime useful.
> 
> When I left `docker-compose` in Makefile, it just said that `docker-compose` is not a command and the whole `make up` failed immediately. After I changed it to `podman-compose` it started doing something.

I deleted the service_health conditions because docker-compose on Podman doesn't support it yet. Other than that, everything worked completely fine on my test as in the ss below.

![2021-08-24_21-29](https://user-images.githubusercontent.com/39852038/130672123-d8e393b8-c3e1-416f-a8a3-030ea3efc332.png)

Zlopez commented:
> I deleted the service_health conditions because docker-compose on Podman doesn't support it yet. Other than that, everything worked completely fine on my test as in the ss below.
> 
> ![2021-08-24_22-21](https://user-images.githubusercontent.com/39852038/130677142-d145633e-4278-4d7a-b588-94414a9a4e55.png)

What do you mean by `docker-compose` on Podman? I just installed `podman-compose` on Fedora.

abitrolly commented:
@Zlopez https://fedoramagazine.org/use-docker-compose-with-podman-to-orchestrate-containers-on-fedora/

Zlopez commented:
Now I understand, I need `docker-compose` installed as well.

ozkanonur commented:
Also, here you can see about podman with docker-compose [this](https://www.redhat.com/sysadmin/podman-docker-compose)

Zlopez commented:
I read both of those, but I understood that you only need to set alias for `docker-compose` to make this work. Maybe I didn't read them carefully enough.

ozkanonur commented:
You just need to install `podman-docker` and `docker-compose`, also make sure the Podman socket is running. Then you should able to use docker-compose without any issue.

ozkanonur commented:
Podman officially provided this compatibility, so I thought it's better to go the official way rather than using 3rd party solution like `podman-compose`.

Zlopez commented:
I didn't know that the `podman-compose` is 3rd party solution. I thought it's the official thing from podman. In this case `docker-compose` with `podman-docker` sounds like a better way to go.

ozkanonur commented:
> I didn't know that the `podman-compose` is 3rd party solution. I thought it's the official thing from podman. In this case `docker-compose` with `podman-docker` sounds like a better way to go.

Looks like they moved the `podman-compose` under the 'Containers' organization. It was a personal project of someone before. We can also use podman-compose if you want. But I recommend sticking with docker-compose since it's supported by podman. That way we can simply use docker-compose without checking should we run docker?podman-compose in some scripts in order to execute commands.

Zlopez commented:
> > I didn't know that the `podman-compose` is 3rd party solution. I thought it's the official thing from podman. In this case `docker-compose` with `podman-docker` sounds like a better way to go.
> 
> Looks like they moved the `podman-compose` under the 'Containers' organization. It was a personal project of someone before. We can also use podman-compose if you want. But I recommend sticking with docker-compose since it's supported by podman. That way we can simply use docker-compose without checking should we run docker?podman-compose in some scripts in order to execute commands.
> 
> Those thoughts I wrote above are my personal thoughts. Both ways are fine for me to develop. If you say "podman-compose is better to go", then I can refactor the scripts for it.

Let's go with `docker-compose`. As you say, in this way we don't need to force people to use podman, if they already have docker on their machine.

ozkanonur commented:
![2021-09-01_18-29](https://user-images.githubusercontent.com/39852038/131699895-9118edbc-cf1d-4bb8-ad34-f9a666a1a9d0.png)

For every compose command, this question will pop up. Is this usage okay ? @abitrolly @Zlopez 

The other option is providing environment variable. For example,

``COMPOSE_TOOL=docker-compose``
or
``COMPOSE_TOOL=podman-compose``

abitrolly commented:
@ozkanonur just do autodetection. Actually, I would prefer `podman play` if `podman` is detected, and I thought that `podman generate` can be used to convert `docker-compose.yml` to Kubernetes format, but `podman generate -h` doesn't confirm it is possible.

abitrolly commented:
Opened an issue for making `podman` more friendly to `docker-compose` users https://github.com/containers/podman/issues/11394

ozkanonur commented:
> @ozkanonur just do [autodetection](https://stackoverflow.com/questions/592620/how-can-i-check-if-a-program-exists-from-a-bash-script). Actually, I would prefer `podman play` if `podman` is detected, and I thought that `podman generate` can be used to convert `docker-compose.yml` to Kubernetes format, but `podman generate -h` doesn't confirm it is possible.

I thought maybe someone will have podman but still wants to keep docker-compose. Now, I provided the conditional compose execution. So you can test it.

Zlopez commented:
I tried to test it again today.

1. `make up` doesn't work for me for rabbitmq and postgres. Getting 404 when selecting the source of image (tried fedoraproject.org and quay.io) See [container.log](https://github.com/fedora-infra/anitya/files/7097611/container.log) for more info

2. `make bash-web` fails
```
podman-compose -f container-compose.yml exec anitya-web bash -c "cat /app/ansible/roles/anitya-dev/files/motd; bash;"
usage: podman-compose [-h] [-f file] [-p PROJECT_NAME] [--podman-path PODMAN_PATH] [--no-ansi] [--no-cleanup] [--dry-run] [-t {1pod,1podfw,hostnet,cntnet,publishall,identity}]
                      {help,version,pull,push,build,up,down,ps,run,start,stop,restart,logs} ...
podman-compose: error: argument command: invalid choice: 'exec' (choose from 'help', 'version', 'pull', 'push', 'build', 'up', 'down', 'ps', 'run', 'start', 'stop', 'restart', 'logs')
make: *** [Makefile:21: bash-web] Error 2
```
 When trying to run it manually by calling `podman exec anitya-web bash -c "cat /app/ansible/roles/anitya-dev/files/motd; bash;"` it will just show the MOTD and kicks me out of container.

3. `make bash-consumer` the same as above
4. `make logs` fails
```
podman-compose -f container-compose.yml logs -f
usage: podman-compose logs [-h] [-f] [-t] [--tail TAIL] service
podman-compose logs: error: the following arguments are required: service
make: *** [Makefile:25: logs] Error 2
```

I didn't tried the database related commands in `Makefile`, because the postgres container isn't running.

ozkanonur commented:
> I tried to test it again today.
> 
> 1. `make up` doesn't work for me for rabbitmq and postgres. Getting 404 when selecting the source of image (tried fedoraproject.org and quay.io) See [container.log](https://github.com/fedora-infra/anitya/files/7097611/container.log) for more info
> 2. `make bash-web` fails
> 
> ```
> podman-compose -f container-compose.yml exec anitya-web bash -c "cat /app/ansible/roles/anitya-dev/files/motd; bash;"
> usage: podman-compose [-h] [-f file] [-p PROJECT_NAME] [--podman-path PODMAN_PATH] [--no-ansi] [--no-cleanup] [--dry-run] [-t {1pod,1podfw,hostnet,cntnet,publishall,identity}]
>                       {help,version,pull,push,build,up,down,ps,run,start,stop,restart,logs} ...
> podman-compose: error: argument command: invalid choice: 'exec' (choose from 'help', 'version', 'pull', 'push', 'build', 'up', 'down', 'ps', 'run', 'start', 'stop', 'restart', 'logs')
> make: *** [Makefile:21: bash-web] Error 2
> ```
> 
> When trying to run it manually by calling `podman exec anitya-web bash -c "cat /app/ansible/roles/anitya-dev/files/motd; bash;"` it will just show the MOTD and kicks me out of container.
> 
> 1. `make bash-consumer` the same as above
> 2. `make logs` fails
> 
> ```
> podman-compose -f container-compose.yml logs -f
> usage: podman-compose logs [-h] [-f] [-t] [--tail TAIL] service
> podman-compose logs: error: the following arguments are required: service
> make: *** [Makefile:25: logs] Error 2
> ```
> 
> I didn't tried the database related commands in `Makefile`, because the postgres container isn't running.

Which version of podman and podman-compose are you using ? I want test this with using them.

Zlopez commented:
I'm trying to run the latest versions available on Fedora.

```
$ podman --version
podman version 3.3.0

$ rpm -qa | grep podman-compose
podman-compose-0.1.7-4.git20210129.fc34.noarch
```

abitrolly commented on Containerfile.dev:
```diff
@@ -1,5 +1,5 @@
 ARG FEDORA_VERSION
-FROM fedora:${FEDORA_VERSION}
+FROM docker.io/library/fedora:${FEDORA_VERSION}
```
Why not `registry.fedoraproject.org` then? :)

ozkanonur replied:
I thought maybe dockerhub has the fastest support. But sure we can change it to that.

ozkanonur replied:
What about for postgresql and rabbitmq. Should I change their registry aswell?

Zlopez replied:
I'm not sure if postgresql or rabbitmq could be found on `registry.fedoraproject.org`

ozkanonur replied:
They dont exists under fedora registry. I thought maybe we can use redhat's registry or something.

ozkanonur commented:
@Zlopez When you have time, please test the make commands with purging previous builds. Once this works with no issue, I will refactor hotness containers because its so much different than this.

ozkanonur commented on docs/contributing.rst:
```diff
@@ -181,6 +181,66 @@ To run check service simply run::
+     - anitya:anitya
+   * - PostgreSQL
+     - http://localhost:5432
+     - postgres:anypasswordworkslocally
```
What exactly do you mean by making services local to pod? Do you mean port forwarding?

Zlopez commented:
@ozkanonur I tested it again. Here is the output:

1. `make up` works without issue
2. `make halt` works without issue
3. `make restore-dump` works
4. `make restart` fails
```
podman-compose -f container-compose.yml restart
usage: podman-compose restart [-h] [-t TIMEOUT] services [services ...]
podman-compose restart: error: the following arguments are required: services
make: *** [Makefile:21: restart] Error 2
```
5. `make bash-web` works, but the MOTD needs to be completely rewritten, `systemd` doesn't work in container and I noticed that Anitya isn't running by default
6. `make bash-consumer` same as above

ozkanonur commented:
> @ozkanonur I tested it again. Here is the output:
> 
> 1. `make up` works without issue
> 2. `make halt` works without issue
> 3. `make restore-dump` works
> 4. `make restart` fails
> 
> ```
> podman-compose -f container-compose.yml restart
> usage: podman-compose restart [-h] [-t TIMEOUT] services [services ...]
> podman-compose restart: error: the following arguments are required: services
> make: *** [Makefile:21: restart] Error 2
> ```
> 
> 1. `make bash-web` works, but the MOTD needs to be completely rewritten, `systemd` doesn't work in container and I noticed that Anitya isn't running by default
> 2. `make bash-consumer` same as above

4- should be fixed now.

For 5 and 6 I will rewrite the MOTD today later.

ozkanonur commented:
New motd files are created for consumer and web pods.

ozkanonur commented:
May I know the state of this PR ? @abitrolly @Zlopez

abitrolly commented:
Well, I don't know how to check if Zuul tested Anitya in containers, and how its coverage is different from normal Ansible setup. Right now I don't need anything to be fixed it Anitya, so it is unlikely that I will test it soon.

The `.container` convention is quite unusual. I haven't seen it before. Did you follow some guidelines for that? I can't say that I like  the practice of placing container run-time volumes into git project subdirectories.

ozkanonur commented:
> Well, I don't know how to check if Zuul tested Anitya in containers, and how its coverage is different from normal Ansible setup. Right now I don't need anything to be fixed it Anitya, so it is unlikely that I will test it soon.
> 
> The `.container` convention is quite unusual. I haven't seen it before. Did you follow some guidelines for that? I can't say that I like the practice of placing container run-time volumes into git project subdirectories.

It's usually called `.docker`. Since we work for both docker and podman, I decided to rename it as `.container`.

Zlopez commented:
@ozkanonur I thought it's still work in progress.

abitrolly commented:
> It's usually called .docker. Since we work for both docker and podman, I decided to rename it as .container.

Do you have a reference where this convention comes from?

ozkanonur commented:
> > It's usually called .docker. Since we work for both docker and podman, I decided to rename it as .container.
> 
> Do you have a reference where this convention comes from?

I don't think there is an official article that refers to it. It's just like the other dot directories where you keep your config and temporary files and folders. In our case, we are keeping DB and motd files for containers. Basically, instead of using separated directories in the repository, you keep them all in a single directory and manage all the container configuration and persistence files/folders by a single volume.

Here you can see this technique has been used by people and docker's organization.

https://github.com/docker/docker.github.io
https://github.com/yeasy/docker_practice
https://github.com/nurdism/neko

Zlopez commented on Makefile:
```diff
@@ -0,0 +1,35 @@
+	$(call container-tool) exec -it anitya-web bash -c "cat /app/.container/web/motd; bash;"
+bash-consumer:
+	$(call container-tool) exec -it anitya-librariesio-consumer bash -c "cat /app/.container/consumer/motd; bash;"
+init-db:
```
This should be run automatically when the container is created. You can't even login to Anitya if the alembic migrations are not run, it's missing the socialauth tables.

Zlopez reviewed (changes requested):
Found few things that should be addressed, but all the make commands are working fine.

Zlopez commented:
@ozkanonur Sorry didn't noticed the your comment. The tox has indeed it's own environment for running, but the psycopg2 needs some libraries that needs to be installed on the testing machine itself, unfortunately.

abitrolly commented:
I gave up on dependencies and followed the Ansible route https://github.com/fedora-infra/anitya/commit/dcdc94169525b515dfb13de06f4ec1a9f7fd7479

abitrolly commented:
But standard practice is to install `psycopg2-binary` if you speicifically need PostreSQL bindings for development.

ozkanonur commented:
> @ozkanonur Sorry didn't noticed the your comment. The tox has indeed it's own environment for running, but the psycopg2 needs some libraries that needs to be installed on the testing machine itself, unfortunately.

So when psycopg2 is provided in tox environment, tox command works in containers but py38,39 and diff-cover jobs fails. When psycopg2 is not provided in tox environment then it fails in containers but doesn't fail on zuul.

ozkanonur commented:
> But standard practice is to install `psycopg2-binary` if you speicifically need PostreSQL bindings for development.

Oh true I still didn't try that one, let me try that quickly.

ozkanonur commented:
Works in containers, let's see if Zuul jobs will pass

ozkanonur commented:
> But standard practice is to install `psycopg2-binary` if you speicifically need PostreSQL bindings for development.

Thank you

Zlopez commented:
I tested it today, it seems like the tox is now running for most of the targets. Only target that is not running is `docs` which is failing with:
```
docs run-test-pre: PYTHONHASHSEED='2745212726'
docs run-test: commands[0] | mkdir -p _static
docs run-test: commands[1] | rm -rf _build
docs run-test: commands[2] | mkdir -p images
docs run-test: commands[3] | ./generate_db_schema
[anitya.config INFO] Loading Anitya configuration from /etc/anitya/anitya.toml
[anitya.config WARNING] SECRET_KEY is not configured, falling back to the default. This is NOT safe for production deployments!
/app/.tox/docs/lib/python3.9/site-packages/anitya/lib/versions/rpm.py:49: UserWarning: Failed to import 'rpm', emulating RPM label comparisons
  warnings.warn("Failed to import 'rpm', emulating RPM label comparisons")
docs run-test: commands[4] | sphinx-build -W -b html -d /app/.tox/docs/tmp/doctrees . _build/html
Running Sphinx v4.2.0
[anitya.config INFO] Loading Anitya configuration from /etc/anitya/anitya.toml
[anitya.config WARNING] SECRET_KEY is not configured, falling back to the default. This is NOT safe for production deployments!
/app/.tox/docs/lib/python3.9/site-packages/anitya/lib/versions/rpm.py:49: UserWarning: Failed to import 'rpm', emulating RPM label comparisons
  warnings.warn("Failed to import 'rpm', emulating RPM label comparisons")
making output directory... done
[autosummary] generating autosummary for: admin-guide.rst, admin-user-guide.rst, api.rst, contributing.rst, database.rst, docblocks/anitya.db.models.rst, glossary.rst, index.rst, integrating-with-anitya.rst, release-notes.rst, user-guide.rst
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
loading intersphinx inventory from https://fedora-messaging.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from https://docs.sqlalchemy.org/en/latest/objects.inv...
intersphinx inventory has moved: https://docs.sqlalchemy.org/en/latest/objects.inv -> https://docs.sqlalchemy.org/en/14/objects.inv
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 11 source files that are out of date
updating environment: [new config] 11 added, 0 changed, 0 removed
reading sources... [  9%] admin-guide
reading sources... [ 18%] admin-user-guide
reading sources... [ 27%] api

Exception occurred:
  File "/app/.tox/docs/lib64/python3.9/site-packages/sqlalchemy/dialects/postgresql/psycopg2.py", line 782, in dbapi
    import psycopg2
ModuleNotFoundError: No module named 'psycopg2'
The full traceback has been saved in /tmp/sphinx-err-gx5u2ph0.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

@ozkanonur Does it working for you?

ozkanonur commented:
> I tested it today, it seems like the tox is now running for most of the targets. Only target that is not running is `docs` which is failing with:
> 
> ```
> docs run-test-pre: PYTHONHASHSEED='2745212726'
> docs run-test: commands[0] | mkdir -p _static
> docs run-test: commands[1] | rm -rf _build
> docs run-test: commands[2] | mkdir -p images
> docs run-test: commands[3] | ./generate_db_schema
> [anitya.config INFO] Loading Anitya configuration from /etc/anitya/anitya.toml
> [anitya.config WARNING] SECRET_KEY is not configured, falling back to the default. This is NOT safe for production deployments!
> /app/.tox/docs/lib/python3.9/site-packages/anitya/lib/versions/rpm.py:49: UserWarning: Failed to import 'rpm', emulating RPM label comparisons
>   warnings.warn("Failed to import 'rpm', emulating RPM label comparisons")
> docs run-test: commands[4] | sphinx-build -W -b html -d /app/.tox/docs/tmp/doctrees . _build/html
> Running Sphinx v4.2.0
> [anitya.config INFO] Loading Anitya configuration from /etc/anitya/anitya.toml
> [anitya.config WARNING] SECRET_KEY is not configured, falling back to the default. This is NOT safe for production deployments!
> /app/.tox/docs/lib/python3.9/site-packages/anitya/lib/versions/rpm.py:49: UserWarning: Failed to import 'rpm', emulating RPM label comparisons
>   warnings.warn("Failed to import 'rpm', emulating RPM label comparisons")
> making output directory... done
> [autosummary] generating autosummary for: admin-guide.rst, admin-user-guide.rst, api.rst, contributing.rst, database.rst, docblocks/anitya.db.models.rst, glossary.rst, index.rst, integrating-with-anitya.rst, release-notes.rst, user-guide.rst
> loading intersphinx inventory from https://docs.python.org/3/objects.inv...
> loading intersphinx inventory from https://fedora-messaging.readthedocs.io/en/latest/objects.inv...
> loading intersphinx inventory from https://docs.sqlalchemy.org/en/latest/objects.inv...
> intersphinx inventory has moved: https://docs.sqlalchemy.org/en/latest/objects.inv -> https://docs.sqlalchemy.org/en/14/objects.inv
> building [mo]: targets for 0 po files that are out of date
> building [html]: targets for 11 source files that are out of date
> updating environment: [new config] 11 added, 0 changed, 0 removed
> reading sources... [  9%] admin-guide
> reading sources... [ 18%] admin-user-guide
> reading sources... [ 27%] api
> 
> Exception occurred:
>   File "/app/.tox/docs/lib64/python3.9/site-packages/sqlalchemy/dialects/postgresql/psycopg2.py", line 782, in dbapi
>     import psycopg2
> ModuleNotFoundError: No module named 'psycopg2'
> The full traceback has been saved in /tmp/sphinx-err-gx5u2ph0.log, if you want to report the issue to the developers.
> Please also report this if it was a user error, so that a better error message can be provided next time.
> A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
> ```
> 
> @ozkanonur Does it working for you?

It didn't work. I had to freshly build the containers to see this problem. Now all works without any issue.

Zlopez commented:
@ozkanonur It looks like we need to manually rebase to latest master.

abitrolly commented:
Looking at the terminal commands, I believe there were some conventions in big companies on common commands that are used by people to build and test projects, so that the project don't have to write readmes with all the same info. Wish I could remember who it was.

Zlopez commented:
@abitrolly I never heard about official conventions for common commands. But I believe that there is probably some.

abitrolly commented:
Found them.

https://github.com/github/scripts-to-rule-them-all

ozkanonur commented:
I wanted to provide similar commands to the existing vagrant commands. We can change them easy if necessary.
